### PR TITLE
packed encoding deserialization can now be turned off

### LINKED ISSUE #112: What if I don't want the packed serialization?
I'm trying to avoid having a format that inadvertently works. While avoiding serialization is a matter of just not calling the `_packed` functions, deserialization seems to implicitly always use it.

E.g., coding against a spec, I do not want other, incorrect (according to the spec, which might be agnostic to Rust/`serde_cbor`) to successfully deserialize, lest someone thinks that format *is* supported.

### PR DESCRIPTION
**Pull request by baloo**
Fixes #112 
See #119

### CODE REVIEW

**Comment by pyfisch** [comment]
looks like it needs a `cargo fmt --all`

**Comment by baloo** [comment]
Ha! I didn't know the tests were not formated by `cargo fmt`! Thanks!

**Comment by baloo** [comment]
I also added tests for the indefinite length map (`0xbf`)

**Comment by pyfisch** [comment]
LGTM.

For symmetry I also want to support the `accept_named` option (to opt out of named fields). Would you also like to implement this here?

File: src/de.rs [hunk 1]
```diff
@@ -947,6 +969,16 @@ where
         }
         *self.len -= 1;
 
+        match self.de.peek()? {
+            Some(_byte @ 0x00..=0x1b) if !self.accept_packed => {
+                return Err(self.de.error(ErrorCode::WrongStructFormat));
+            }
+            Some(_byte @ 0x40..=0x7f) if !self.accept_named => {
```
Comment (baloo) [inline src/de.rs]: I used range 0x40 which is for byte string here. I felt I needed to mention it.

File: src/de.rs [hunk 2]
```diff
@@ -987,6 +1021,12 @@ where
         K: de::DeserializeSeed<'de>,
     {
         match self.de.peek()? {
+            Some(_byte @ 0x00..=0x1b) if !self.accept_packed => {
+                return Err(self.de.error(ErrorCode::WrongStructFormat))
+            }
+            Some(_byte @ 0x40..=0x7f) if !self.accept_named => {
```
Comment (baloo) [inline src/de.rs]: I used range 0x40 which is for byte string here. I felt I needed to mention it.

**Comment by baloo** [comment]
I moved to a struct to feed arguments to `from_slice_stream_options`. I felt I was loosing expressiveness, and this approach felt more expressive to me. Let me know what you think.

**Comment by pyfisch** [comment]
Using a struct `Options` in the tests is fine with me.

File: src/de.rs (hunk 1 above)
Comment (pyfisch) [reply src/de.rs]: I'm a bit confused, I thought fields would only use UTF-8 names and not binary names?

**Comment by baloo** [comment]
@pyfisch anything else needed here? or can we merge this one?

**Comment by pyfisch** [comment]
> @pyfisch anything else needed here? or can we merge this one?

https://github.com/pyfisch/cbor/pull/126#discussion_r294075703

File: src/de.rs (hunk 1 above)
Comment (baloo) [reply src/de.rs]: ha, missed that one! well I'm not too sure. That's one of the reasons I raised this question.

**Comment by baloo** [comment]
Just pushed a last commit to try to settle the argument. Serde is happily using byte string as struct field names instead of UTF8 Strings.
In my opinion, this should not be accepted if we were to disable unpacked deserialization.

WDYT?

**Comment by pyfisch** [comment]
>  In my opinion, this should not be accepted if we were to disable unpacked deserialization.

I agree. Only UTF-8 strings should be accepted. (I guess this was added by serde to support formats that don't differentiate between text and binary strings.)

**Comment by pyfisch** [comment]
If you remove binary strings I can merge it.

**Comment by baloo** [comment]
@pyfisch There

**Comment by pyfisch** [comment]
:tada:

**Pull request merged**